activation/handler_v2.go

```diff
@@ -248,3 +306,319 @@ func (h *HandlerV2) collectAtxDeps(atx *wire.ActivationTxV2) ([]types.Hash32, []
+	case types.AtxV1:
+		var prev wire.ActivationTxV1
+		if err := codec.Decode(blob.Bytes, &prev); err != nil {
+			return nil, fmt.Errorf("decoding previous atx v1: %w", err)
+		}
+		return &prev, nil
+	case types.AtxV2:
+		var prev wire.ActivationTxV2
+		if err := codec.Decode(blob.Bytes, &prev); err != nil {
+			return nil, fmt.Errorf("decoding previous atx v2: %w", err)
+		}
+		return &prev, nil
+	}
+	return nil, fmt.Errorf("unexpected previous ATX version: %d", version)
+}
+
+// Validate the previous ATX for the given PoST and return the effective numunits.
+func (h *HandlerV2) validatePreviousAtx(id types.NodeID, post *wire.SubPostV2, prevAtxs []opaqueAtx) (uint32, error) {
+	if post.PrevATXIndex > uint32(len(prevAtxs)) {
+		return 0, fmt.Errorf("prevATXIndex out of bounds: %d > %d", post.PrevATXIndex, len(prevAtxs))
+	}
+	prev := prevAtxs[post.PrevATXIndex]
+
+	switch prev := prev.(type) {
+	case *types.ActivationTx:
+		// A golden ATX
+		// TODO: support merged golden ATX
+		if prev.SmesherID != id {
+			return 0, fmt.Errorf("prev golden ATX has different owner: %s (expected %s)", prev.SmesherID, id)
+		}
+		return min(prev.NumUnits, post.NumUnits), nil
+
+	case *wire.ActivationTxV1:
+		if prev.SmesherID != id {
+			return 0, fmt.Errorf("prev ATX V1 has different owner: %s (expected %s)", prev.SmesherID, id)
+		}
+		return min(prev.NumUnits, post.NumUnits), nil
+	case *wire.ActivationTxV2:
+		// TODO: support previous merged-ATX
+
+		// previous is solo ATX
+		if prev.SmesherID == id {
+			return min(prev.NiPosts[0].Posts[0].NumUnits, post.NumUnits), nil
+		}
+		return 0, fmt.Errorf("previous solo ATX V2 has different owner: %s (expected %s)", prev.SmesherID, id)
+	}
+	return 0, fmt.Errorf("unexpected previous ATX type: %T", prev)
+}
+
+// VRF nonce must be valid for the collected space of all included IDs.
+// TODO: support merged ATXs.
+func (h *HandlerV2) validateVrfNonce(
+	atx *wire.ActivationTxV2,
+	previous opaqueAtx,
+	commitment types.ATXID,
+) error {
+	numUnits := atx.TotalNumUnits()
+	needRecheck := numUnits > previous.TotalNumUnits() || atx.VRFNonce != nil
+
+	if atx.VRFNonce == nil {
+		// lookup the previous nonce by this ID
+		// TODO: support merged ATXs
+		// For a merged ATX we need to follow ATXs chain for `id`, find the last ATX by this ID
+		// and check the nonce from it.
+		if atx.MarriageATX != nil {
+			return errors.New("merged ATXs are not supported")
+		}
+		nonce, err := atxs.NonceByID(h.cdb, previous.ID())
+		if err != nil {
+			return fmt.Errorf("getting nonce from previous ATX %s: %w", previous.ID(), err)
+		}
+		atx.VRFNonce = (*uint64)(&nonce)
+	}
+
+	if needRecheck {
+		return h.nipostValidator.VRFNonceV2(atx.SmesherID, commitment, types.VRFPostIndex(*atx.VRFNonce), numUnits)
+	}
+	return nil
+}
+
+func (h *HandlerV2) validateCommitmentAtx(golden, commitmentAtxId types.ATXID, publish types.EpochID) error {
+	if commitmentAtxId != golden {
+		commitment, err := atxs.Get(h.cdb, commitmentAtxId)
```

Comment:
Querying the full set of ATX fields is not very efficient. For one thing, if we don't need the blob of the commitment ATX here, and doing JOIN on the `atx_blobs` table (which `atxs.Get` does) will slow down validation process. Another thing to consider is that multiple separate SQL queries are always slower than a single query, even if the single query fetches some excess fields. Maybe we could consider doing a single query for the deps along the lines of
```sql
select epoch, nonce, num_units where id in (?, ?, ?)
```
using IDs of commitment / prev / positioning ATXs?

Reply:
>For one thing, if we don't need the blob of the commitment ATX here, and doing JOIN on the atx_blobs table (which atxs.Get does) will slow down validation process. 

The `atxs.Get()` doesn't read the blob anymore. It only reads the `atxs` table.

Reply:
I'm not sure a single query is easily doable during the handling of an incoming ATXs, since there are so many branches that need to be considered depending on

- initial vs. non-initial
- merged vs. not-merged
- ATX that resizes post vs. one that doesn't
- etc.

We could consider doing all in a single TX and see if that decreased the validation time.